### DISCUSSION TITLE
UnetStack Community edition for ARM-based Linux SBCs

### DISCUSSION THREAD

**Discussion by lucvt001** [org-arl/unet-support#5] [Q&A] [answered]
Hi Unet Team,

First of all I would like to say thank you for providing making this wonderful tool free for everyone to use. It is tremendously helpful, especially to students and researchers.

At Mecatron, NTU, we are planning to deploy UnetStack on Jetson Nano (arm-based, ubuntu 18.04) and Jetson Orin Nano (arm-based, ubuntu 20.04) with the guidance provided by the DIY acoustic modem blog, to enable two AUVs to communicate in the water. As I understand, the community versions are either for x86_64 Linux, or for Raspberry Pi when it comes to embedded devices. 

May I ask if there are download versions for the two arm-based ubuntu devices that I mention above?

Any help or guidance is much appreciated!

**Comment by notthetup** [comment]
Most of UnetStack3 is written in Java and targets JRE8. JRE is available for Linux on ARM64 from many vendors, e.g. https://adoptium.net/temurin/releases/?version=8&arch=aarch64 . This should work on any Jetson boards.

When using "Unet audio" in UnetStack3, the UnetStack connects to the platform's audio interface through some native code. Those bindings and some other low-level code are included in Unet Community builds as a native library `libyoda_phy_pa*.so/dylib`.

Currently the [unet-community-3.4.4 release](https://unetstack.net/downloads/unet-community-3.4.4.tgz) supports the following platforms

- `macOS/x86`
- `macOS/aarch64`
- `Linux/x86`

The version from the [DIY modem blog post](https://github.com/org-arl/unet-doc/raw/master/docs/downloads/unet-community-pi-3.1.0.tgz) is slightly older and only contains the support for 

- `Linux/armhf`

Technically, there's nothing in the native library that binds it to a specific hardware or libraries. The main dependencies are `glibc` and `portaudio`. So nothing should stop you from running the version from the blog post on a Jetson board or any ARM-based board running Linux.

I agree that it would be slightly simpler and more performant if a version compiled for `aarch64` instead of `armhf` were available. Let me look into what would be needed to create that.

**Reply by TruongQuan777** [reply]
Hello sir,

Thanks for your information. I am a member of NTU's mecatron team and when I try running Unetsstack on our Jetson Nano, several issues appear. The native library libyoda_phy_pa_aarch64 was not found as it was in .dylib type, so when I try to rename the libyoda_phy_pa_arrch64 from .dylib into .so file and run, the ELF header problem appear. 

Previously, the java version for armhf  we downloaded into our Jetson also rise similiar issues so it also appear that using the armhf version on Jetson Nano also does not work.

May I ask if you have any suggestion or a compiled version for aarch64?

Thanks for your assistance.

**Reply by notthetup** [reply]
@TruongQuan777 

A `.dylib` file is typically a [Mach-O object file (shared library) for macOS/OSX](https://en.wikipedia.org/wiki/Mach-O). A `.so` file is the equivalent in Linux. (For completeness, a `.dll` is the Windows equivalent).

You're probably running Linux on your Jetson Nano. So, Java will look for a `.so` file when it tries to load the native code library. However as I mentioned above, [unet-community-3.4.4 release](https://unetstack.net/downloads/unet-community-3.4.4.tgz) doesn't support Linux on `aarch64`/`armhf`. So none of THOSE libraries would work for you.

```
[7:53]chinmay@gauss:~ ❯ tar -tf /Users/chinmay/Downloads/unet-community-3.4.4.tgz | grep libyoda
unet-3.4.4/lib/libyoda_phy_pa_aarch64.dylib
unet-3.4.4/lib/libyoda_phy_pa_amd64.so
unet-3.4.4/lib/libyoda_phy_pa_x86_64.dylib
```

The version from the [DIY modem blog post](https://github.com/org-arl/unet-doc/raw/master/docs/downloads/unet-community-pi-3.1.0.tgz) has support for `Linux/armhf` so it should have a `libyoda_phy_pa_arm.so` file.

```
[7:53]chinmay@gauss:~ ❯ tar -tf /Users/chinmay/Downloads/unet-community-pi-3.1.0.tgz | grep libyoda
unet-3.1.0/lib/libyoda_phy_pa_arm.so
```

You might need to rename the file to `libyoda_phy_pa_aarch64.so` or something else depending on what the JVM you're using refers to for the platform architecture (`System.properties['os.arch']`).  

The JNI code looks for a native library file with the name format `'libyoda_phy_pa_'+System.properties['os.arch']` and an OS-appropriate extension.

**Reply by TruongQuan777** [reply, marked as answer]
Thanks for your suggestion sir. Also really sorry for the late response.

We were able to run the Unetstack on the Jetson Nano. At first we try the method you suggest but it seems like the Jetson still does not recognize the file. So what we did was to still use the version you suggest but keep the file name as it is (libyoda_phy_pa_armhf.so instead of libyoda_phy_pa_aarch64.so) and we had to install the 32-bit Java instead of 64-bit Java to run that file. 

However, after successfully running the Unetstack it seems like the Unetstack doesn't really work with our external soundcard (the speaker still make sound when I watch youtube but it doesn't have any sound when I run Unetstack). I think we can solve this problem quite fast so I will update to you when its done

**Reply by xdead10cc** [reply]
Dear @TruongQuan777 -- I wonder if you were ever able to resolve the sound problem. I have reached to the same point and now I am finding myself short on ideas. Kindly share if you managed to generate audio using Unetstack.

Thanks !

**Reply by TruongQuan777** [reply]
Hi @xdead10cc yes I was able to solve. The solution is simply open the asoundsrc file. Then, in the asoundsrc file, you change the line  "pcm !. default" and the "pcm device" (or something like that) into the right card and device number of your speaker ( you can find that card and device number using the aplay-l command). For example, on my site, I change it to card 2, device 0.

**Reply by TruongQuan777** [reply]
Dear @notthetup, after installing the 32-bit Java to open the Unetstack and changing the card and device number in our asoundrc.conf file as we mentioned above, we were able to run unetstack and emit sound from our speaker. 

However, one new problem we face is that when we use our  Jetson Nano (using an external speaker) to send the message to my laptop, I can only send it for the first time. After that, when sending there will be no more sound and the laptop wont receive it.

To resolve it we have to close the terminal on the Jetson Nano, open the Unetstack again everytime we send a new message. Do you have any suggestion for solving this issue? Thanks!

**Reply by notthetup** [reply]
@TruongQuan777 That's strange. Could you open a new discussion about this issue, with all the details of your setup and any log files you can capture? It'd be more useful for folks facing a similar problem.

**Comment by lucvt001** [comment]
Hi @notthetup thank you for your fantastic response! On our side, we will try unet-3.4.4 on the Jetson boards and update you if we manage to fix all bugs.
